File: examples/selenium-java-03-multifile-login/input/pages/LoginPage.java
package com.acme.shop.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.How;
import org.openqa.selenium.support.ui.ExpectedConditions;

public class LoginPage extends BasePage {

    private static final String URL = "https://shop.acme.test/login";

    @FindBy(id = "email") private WebElement emailInput;
    @FindBy(id = "password") private WebElement passwordInput;
    @FindBy(how = How.XPATH, using = "//form//button[@type='submit']") private WebElement signInButton;
    @FindBy(css = ".error-banner") private WebElement errorBanner;

    public void open() {
        driver.get(URL);
    }

    public void signIn(String email, String password) throws InterruptedException {
        emailInput.clear();
        emailInput.sendKeys(email);
        passwordInput.clear();
        passwordInput.sendKeys(password);
        signInButton.click();
        Thread.sleep(1500);
    }

    public boolean isOnDashboard() {
        return wait.until(ExpectedConditions.urlContains("/dashboard"));
    }

    public String errorText() {
        waitVisible(errorBanner);
        return errorBanner.getText();
    }

    public boolean hasError() {
        return isVisibleSafe(By.cssSelector(".error-banner"));
    }
}

File: examples/selenium-python-03-multifile-login/input/pages/login_page.py
"""LoginPage — exercises common Selenium-Python smells.

- WebDriverWait + ExpectedConditions ceremony
- find_element with positional XPath
- page_source substring assertion for dashboard check
- assertion via raw element.text equality
"""

import time

from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import WebDriverWait

from pages.base_page import BasePage


class LoginPage(BasePage):
    PATH = "/sign-in"

    EMAIL_INPUT = (By.ID, "email")
    PASSWORD_INPUT = (By.ID, "password")
    SUBMIT_BUTTON = (By.XPATH, "//form//button[@type='submit']")
    ERROR_MESSAGE = (By.CSS_SELECTOR, ".form-error")
    DASHBOARD_HEADER = (By.XPATH, "//header//h1[contains(text(), 'Dashboard')]")

    def sign_in(self, email, password):
        self.driver.find_element(*self.EMAIL_INPUT).send_keys(email)
        self.driver.find_element(*self.PASSWORD_INPUT).send_keys(password)
        self.driver.find_element(*self.SUBMIT_BUTTON).click()
        time.sleep(1)

    def is_on_dashboard(self):
        try:
            WebDriverWait(self.driver, 5).until(
                EC.presence_of_element_located(self.DASHBOARD_HEADER)
            )
            return True
        except Exception:
            return False

    def error_message_text(self):
        return self.driver.find_element(*self.ERROR_MESSAGE).text
